### ISSUE TITLE
Merge bot should check for flutter-gold if `framework` shards were executed

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#95355]
Similar to https://github.com/flutter/flutter/issues/81852

https://github.com/flutter/flutter/pull/95302 was submitted even though it had untriaged golden images. I would have expected that the `flutter-gold` check would have prevented the submission of that PR. But for some reason, that check didn't run on this PR.

The flutter-gold check had not been applied to this PR, so the merge bot landed it.
The merge bot should make sure any change that executes golden file tests (any shard with `framework` in the title is an indicator) has a flutter-gold check to prevent breaking the tree like this again.

**Comment by Piinks** [comment]
A flutter-gold status had not been applied, so it was not there to block the merge bot from landing the change. I'll update this to add that assertion to the merge bot.

**Comment by Piinks** [comment]
Updated OP, similar to https://github.com/flutter/flutter/issues/81852 I think the merge bot should check for this.

**Comment by godofredoc** [comment]
Does this test deterministically run on all PRs?

**Comment by godofredoc** [comment]
A good way to prevent this from happening is to add the flutter-gold check as soon as the PR is created. The additional benefit of this is flutter gold functionality is encapsulated in a single place and the bot won't need specific functionality to deal with this specific use case.

**Comment by Piinks** [comment]
> Does this test deterministically run on all PRs?

Nope, only those with the name `framework` among the shards that are run.

> A good way to prevent this from happening is to add the flutter-gold check as soon as the PR is created.

It can't be applied to every PR because not every PR runs golden file tests. We used to just apply it to all PRs and if they did not execute golden file tests it would just automatically turn green, but feedback was that the check just shouldn't be applied at all if no golden file test are executing.

Since flutter-gold is based on what checks are running, it can't be applied when the PR is opened. :( It has to wait for checks to be scheduled first so it knows whether or not it is relevant.

It is currently applied during the push-gold-status-to-github cron job that runs every 5 minutes, when that job runs, it cycles through every PR to apply flutter-gold. There isn't error reporting for when this cron job fails (https://github.com/flutter/flutter/issues/92503).

So - for example - if the cron job runs into a problem with PR 5, it will never reach PR 10.
PR 10 gets no flutter-gold status and the merge bot lands it like the above referenced engine roll.

I think this will be very helpful in letting us know too when that cron job has an issue.

**Comment by godofredoc** [comment]
From the comments it seems like the problem is the cron job is not reliable. Should we fix that instead of trying to add additional logic to the auto-submit bot? for reference we are not adding any additional logic to handle google testing all the efforts are focused on making the google testing check more reliable and deterministic.

**Comment by godofredoc** [comment]
\cc @CaseyHillers

**Comment by CaseyHillers** [comment]
Would it make sense to integrate this into the ci.yaml logic?

We could add a basic LUCI builder that polls for this, or we can add it as an external test - https://github.com/flutter/cocoon/blob/main/CI_YAML.md#external-tests

**Comment by CaseyHillers** [comment]
Infra is working to generalize the auto-submit logic, and we shouldn't add any edge cases (flutter.dev/go/auto-submit)

Can flutter-gold status be created on every PR immediately, but starting it is deferred until all the other checks are green?

**Comment by Piinks** [comment]
> Can flutter-gold status be created on every PR immediately, but starting it is deferred until all the other checks are green?

We already do this. The flutter-gold check is a cron job that runs every 5 minutes. It cycles through every open PR (in sequential order) and automatically adds a pending flutter-gold check to any PR that executes a `framework` test shard. Once all other tests are complete, it just makes one request to the Gold server to see if there are new images. That's it.

The trouble is, if the flutter-gold check breaks, or gets stuck for one reason or another, and it never applies that initial status, the merge bot can't be blocked by it. So the merge bot will merge PRs that could be introducing new images, which then breaks the tree.

Currently we only use the flutter-gold check on some PRs, not all of them, so I don't know if that already makes the flutter-gold check an edge case... Contributors complained it was irrelevant to have on PRs that do not execute framework tests and was confusing or delaying when waiting 5 minutes for the next job to run.

**Comment by godofredoc** [comment]
Closing because the bug has not been updated for more than one year.

Feel free to reopen any bugs that you believe are important but please provide updates on how to resource the fix of the bug and what can be done to prevent the issue from not getting updates for another year.
